### PR TITLE
Enhancement: Allowing to write rule decorator

### PR DESCRIPTION
**Pull request by smatsumt**
# What's this PR?

This PR allows to write **decorator wrapping a rule** in Snakefile.

This is example of decorator in Snakefile. We can enable remote file access just putting `@s3_remote_wrapper` decorator.
```
@s3_remote_wrapper
rule rule1:
    input: 'some-bucket-name/input.txt'
    output: 'some-bucket-name/output.txt'
    shell: 'grep foo {input} > {output}'
```

Decorator implementation is like this.

```
from snakemake.remote.S3 import RemoteProvider as S3RemoteProvider
S3 = S3RemoteProvider()

def s3_remote_wrapper(ruleinfo):
    # wrap S3.remote() with all args and kwargs of input/output
    ruleinfo.input = (tuple(map(S3.remote, ruleinfo.input[0])), {k: S3.remote(v) for k, v in ruleinfo.input[1]})
    ruleinfo.output = (tuple(map(S3.remote, ruleinfo.output[0])), {k: S3.remote(v) for k, v in ruleinfo.output[1]})
    return ruleinfo
```

# What's the benefit?

- **Keep rule description simple.** Just putting one decorator is simpler than writing `S3.remote()` call for all inputs and outputs.
- **Control wrap functions easily.** The decorator can be easily disabled with just uncomment it. In the example above, we can easily change using remote file or not.
- **Enables to write ruleinfo based behavior.** For example, we can write output filename based on rule name using the ruleinfo argument. This might solve the Issue #156, "Access the rule name within a rule". 

# (Current) Limitation

1. `shellcmd`, `wrapper` and `script` cannot be overwritten. `ruleinfo.shellcmd` is just information, not actual executed command. Assigning `ruleinfo.shellcmd` with new value does not change actual executed command.
2. Getting rulename from ruleinfo is not straight forward. `ruleinfo.func.__name__[7:]`will return rule name.
3. Have to know ruleinfo structure details when writing wrapper.

About limitation 1. and 2., I'm planning to write PR for them if this decorator PR is accepted.

### CODE REVIEW

File: snakemake/parser.py [hunk 1]
```diff
@@ -804,7 +804,7 @@ def __exit__(self, *args):
 def format_tokens(tokens):
     t_ = None
     for t in tokens:
-        if t_ and not t.isspace() and not t_.isspace():
+        if t_ and (not t.isspace() and not t_.isspace()) and (not t_ == '@'):
```
Comment (smatsumt) [inline snakemake/parser.py]: This change is needed to treat decorator correctly.
Before this change, parser changes "@decorator" to "@ decorator", and Python interpreter does not treat it as decorator.

**Comment by johanneskoester** [comment]
Thanks for the suggestion! At the moment, I don't see a use case yet. To shift all IO to e.g. S3, you can simply use --default-remote-provider at the command line. This is preferable, because it keeps platform specific choices out of the workflow definition making it more reproducible for others. For partially having S3 IO, the current way of expressing it is more readable and explicit to non-python experts. Since readability is more important than brevity, I would like to stick with what we have now.

I am always feeling bad about rejecting work, so please accept my deepest apologies. But I also feel responsible to ensure that the snakemake language does not become too heterogeneous and stays readable for everybody.

**Comment by johanneskoester** [comment]
If you disagree or have further arguments, please feel free to open an issue and we can go on with the discussion over there.

**Comment by smatsumt** [comment]
Thank you very much for kind and thoughtful reply.

I agree with that decorator might damage workflow readability. Especially it is true for unfamiliar people. And I understand that "readability is more important than brevity" is important policy for snakemake.

I didn't know that the option `--default-remote-provider` can use for it. Absolutely, it will solve the example usage situation 😃

For me, snakemake is the best workflow engine (that's why I sent the PR 😁 ), I feel the command line options are a little bit too much. I sometimes feel it is hard to find useful options. I guess this comes from that command line options includes both workflow behavior changing options (--force-all, etc) and each task behavior changing option (--default-remote-provider, etc)

I'm thinking second group options might be separated somehow, and making decorator might mitigate it. Actually, I'm not sure decorator is the best solution for it...

Anyway, thanks again for your reply!

**Pull request closed without merging**